src/main/java/org/commcare/session/SessionFrame.java

```diff
@@ -136,15 +136,19 @@ public StackFrameStep popStep() {
         return recentPop;
     }
 
-    protected boolean rewindToMarkAndSet(StackFrameStep step, EvaluationContext evalContext) {
+    protected boolean rewindToMarkAndSet(StackFrameStep step, EvaluationContext evalContext,
+            StackObserver observer) {
         int markIndex = getLatestMarkPosition(steps);
 
         if (markIndex >= 0) {
             String markDatumId = steps.get(markIndex).getId();
+            observer.stepsRewound(steps.subList(markIndex, steps.size()));
             steps = new Vector<>(steps.subList(0, markIndex));
             if (step.getValue() != null) {
                 String evaluatedStepValue = step.evaluateValue(evalContext);
-                steps.addElement(new StackFrameStep(SessionFrame.STATE_UNKNOWN, markDatumId, evaluatedStepValue));
+                StackFrameStep rewindStep = new StackFrameStep(SessionFrame.STATE_UNKNOWN, markDatumId, evaluatedStepValue);
+                steps.addElement(rewindStep);
```

Comment:
Noticed this `addElement` is the same as calling `pushStep` - should the `observer.stepPushed` move into `pushStep`, and this code call that function instead of directly updating `steps`?

Reply:
Similar question about `popStep` - should it be using an observer? Or should it not, because we only need to observe particular code paths?

Reply:
This only instruments changes that happen as a result of stack operations. It could be expanded but we don't have a use case for that and generally those events have clearer triggers outside the session. The changes due to stack ops are mostly invisible from outside.